### DISCUSSION TITLE
Clarion vs Federated Systems (e.g. Matrix, Mastodon, Diaspora)

### DISCUSSION THREAD

**Discussion by bytemaster** [bytemaster/clarion#12] [General]
There are a lot of "decentralized" social media systems that utilize a Federated Model. From what I can deduce from reading the documentation these include [Mastodon](https://mastodon.social/about) and [Diaspora](https://diasporafoundation.org).

In a federated system a "user" has a "home server" (or hub) and hubs are connected together. These "servers" or "hubs" are small in number because of the technical skills required to operate a full node. 

The most widely known example of a Federated System is email. You can have an account@gmail.com and an account@protonmail.com, but these are two different accounts. Your ability to utilize your email account is subject to the email service provider you have chosen. These service providers are in turn subject to all manner of regulation and/or cancel culture pressure.

Under Clarion "email", you would have an account identified by a public key. This single account could be hosted on many different nodes at the same time Red.com and Blue.com and Friend.com and Family.com as well as YourHomeComputerIP.

Someone who wants to send you an email can deliver it to any node operated by a friend of yours. And even if you lose all of your friends, you can still make new friends and retain your Clarion "email" account.

The goal with Clarion is to make it so easy to "admin" a node that most people don't even realize they are doing admin or running a server. You just download an application, run it, and create or import an account, and then invite friends.

**Comment by jerryfane** [comment]
Users can access their "email" directly from a smartphone to connect to other users' nodes, and send/receive forward messages. But this still requires the existence of nodes/users that manage a server, or am I wrong?

**Reply by bytemaster** [reply]
Someone will have to run processes on computers with public IP addresses.  Ideally these "nodes" don't need to be long-lived or permanent.  To install a PWA someone, on some domain, would have to host it.   But once installed you don't need that domain any more, which means it can probably be hosted via IPFS

**Reply by jonathanbahai** [reply]
Do we really need to use IPs? Use a combinations of private/public key + mac address (ID verified) for packet communications for complete agnostic communication within the network. Since the mechanics you are talking about are more concerned with communications within known networks then this becomes easier to accomplish. The IP becomes more of an accessory at that point for the sake of existing software needing one for legacy.

**Reply by bytemaster** [reply]
@jonathanbahai IPv6 could eventually be used.  Don't know that you can connect to a MAC address from a browser/websocket.

**Reply by jonathanbahai** [reply]
From what I understand there is already a namespace translating the addressing.. as long as the packets support is in the nodes, you can do it. I know the name space has to have an IP all the same, but at least it is only s fraction of the overall infrastructure. A custom browser can remove the need all together. Chromium needs a new open source home after what Google did. 

If we are going to decentralized, why continue to put all the control of the network in the hands of ARIN and RIP? :)

**Reply by bytemaster** [reply]
@jonathanbahai this is an area that I am not very familiar with. Can you point to a proof of concept for how this can get around NAT traversal issues?

**Reply by vaultec81** [reply]
@jonathanbahai MAC addresses are primarily meant only for the physical layer routing and not the wider internet. There are also privacy concerns with exposing a MAC address to the greater internet. I am also not sure what it meant by "mac address (ID verified)" There is no such thing as being able to verify a MAC address' authenticity. Browsers definitely do not support raw ethernet frames that would be implied when thinking of using only mac addresses to communicate. Aside from this there are much better approaches to this issue. One being link local IPv6 addresses could be used for local in network routing without the need for a DHCP server or router to handle IP assignment. There is also IP routing technologies such as [yggdrasil](https://yggdrasil-network.github.io/) specifically meant for replacing BGP and the control of ARIN/RIP.

**Reply by jonathanbahai** [reply]
> @jonathanbahai MAC addresses are primarily meant only for the physical layer routing and not the wider internet. There are also privacy concerns with exposing a MAC address to the greater internet. I am also not sure what it meant by "mac address (ID verified)" There is no such thing as being able to verify a MAC address' authenticity. Browsers definitely do not support raw ethernet frames that would be implied when thinking of using only mac addresses to communicate. Aside from this there are much better approaches to this issue. One being link local IPv6 addresses could be used for local in network routing without the need for a DHCP server or router to handle IP assignment. There is also IP routing technologies such as [yggdrasil](https://yggdrasil-network.github.io/) specifically meant for replacing BGP and the control of ARIN/RIP.

Yes I agree. My suggestion was based on smaller networks, which seems to be the intent of this design. Smaller networks that can still talk to each other at certain points thus forming a larger one without needing to have awareness of it all the way current ARP/BGP is needed for the internet with massive routing tables end users never deal with. ISPs do.

To be clear about what I meant using MAC addressing it was "private/public key + mac address" as the suggested addressing which would by association of key to device provide a type of proof of identity the network could use as a means of internally routing the relationships.. again on a small scale though. I know MACs can be played with etc, which is why only ones in combination with the private key would be valid addresses that would message. 

Without ARP and IP header and checksum overhead, it could be blazing fast. All depends how important those 2ms are to you though. :) 

@bytemaster sorry no proof of concept.. just my contribution.. useful or not. :)

**Comment by qimiaoguo** [comment]
your data is hosted by your friends' node and yourself, only all node are lost, your data will be lost, all your data is private, and if you want to send some data to your friends or public, you use your friends' public key or a group key to encrypt it, like keybase
how can you connect to your friends' node?

**Reply by jerryfane** [reply]
> how can you connect o your friends' node?

You should be invited and subscribe your "identity" in their node

https://github.com/bytemaster/clarion/discussions/7

**Comment by lukapercic** [comment]
This sounds the same as [Nostr](https://github.com/fiatjaf/nostr). They share the same criticism for federation and accounts work the user relays - any user can use many relays, and relays don't own the account as they do in federation-based networks.

**Reply by bytemaster** [reply]
Thanks for the heads up on this.  Nostr seems to be the most similar approach I have seen so far.

**Reply by aleclarson** [reply]
There's also [CTZN](https://ctznry.com/).

**Reply by bytemaster** [reply]
CTZN uses DHT and looks similar to most other DHT solutions.

**Comment by maxocoin** [comment]
Did you check https://scuttlebutt.nz/ ?
Looks like you have many goals in common

```

Scuttlebutt is a decentralised secure gossip platform.

Like other social platforms, you can send messages to your friends and share posts onto a feed. The cool thing is that the underlying technology here means that messages are passed directly between friends via a peer-to-peer (p2p) gossip protocol.

As a decentralized social network, Scuttlebutt passes the data from friend to friend, without any central server. The data is localised and distributed so it also happens to work offline!

The name, Scuttlebutt, came from sea-slang for gossip. Basically, like a watercooler on a ship, where sailors and pirates go have a yarn.

```

**Comment by friedger** [comment]
Matrix's Pinecone is a hybrid of a federated system and p2p communication and collaboration network. (https://matrix.org/blog/2021/05/06/introducing-the-pinecone-overlay-network)

**Comment by dentm42** [comment]
The ZOT protocol servers are interesting as they allow channel cloning and multiple home servers for the same identity.  Zap, Hubzilla, Mistypark, etc.  That deals with some of the concerns.

Clarion looks really interesting.  Is there provision for multi-homing an identity in Clarion?  Or is identity tied to the device/instance?
